rir/src/ir/Compiler.cpp

```diff
@@ -1073,6 +1073,87 @@ bool compileSpecialCall(Context& ctx, SEXP ast, SEXP fun, SEXP args_) {
     return false;
 }
 
+SEXP findRIRClosure(SEXP sym, SEXP rho) {
+    SEXP fun;
+    while (rho != R_EmptyEnv) {
+        fun = findVarInFrame3(rho, sym, TRUE);
+        if (fun != R_UnboundValue) {
+            if (TYPEOF(fun) == PROMSXP || TYPEOF(fun) == BUILTINSXP ||
+                    TYPEOF(fun) == SPECIALSXP || fun == R_MissingArg)
+                return nullptr;
+            if (isValidClosureSEXP(fun))
+                return fun;
+        }
+        rho = ENCLOS(rho);
+    }
+    return nullptr;
+}
+
+// try to look up a closure
+bool compileWithGuess(Context& ctx, SEXP ast, SEXP fun, SEXP args_) {
+    RList args(args_);
+    CodeStream& cs = ctx.cs();
+
+    SEXP cls = findRIRClosure(fun, R_GlobalEnv);  // or current env of the closure?
+    if (!cls)
+        return false;
+
+    RList formals(FORMALS(cls));
+    if (formals.length() != args.length() || args.length() == 0)
+        return false;
+
+    for (auto farg : formals)
+        if (farg == R_DotsSymbol)
+            return false;
+
+    for (auto a = args.begin(); a != args.end(); ++a) {
```

Comment:
`for (auto a : args) ...`?

Reply:
Need the iterator itself as I look at the TAGs... range based for would just give me the CARs.